core/store/src/cold_storage.rs

```diff
@@ -0,0 +1,255 @@
+use crate::columns::DBKeyType;
+use crate::{DBCol, DBTransaction, Database, Store};
+
+use borsh::BorshDeserialize;
+use near_primitives::types::BlockHeight;
+use std::collections::HashMap;
+use std::io;
+use std::sync::Arc;
+use strum::IntoEnumIterator;
+
+type StoreKey = Vec<u8>;
+type StoreValue = Option<Vec<u8>>;
+type StoreCache = HashMap<(DBCol, StoreKey), StoreValue>;
+
+struct StoreWithCache<'a> {
+    store: &'a Store,
+    cache: StoreCache,
+}
+
+pub fn update_cold_db(
+    cold_db: &Arc<dyn Database>,
+    hot_store: &Store,
+    height: &BlockHeight,
+) -> io::Result<()> {
+    let _span = tracing::debug_span!(target: "store", "update cold db", height = height);
+
+    let mut store_with_cache = StoreWithCache { store: hot_store, cache: StoreCache::new() };
+
+    let key_type_to_keys = get_keys_from_store(&mut store_with_cache, height)?;
+    for col in DBCol::iter() {
+        if col.is_cold() {
+            copy_from_store(
+                cold_db,
+                &mut store_with_cache,
+                col,
+                combine_keys(&key_type_to_keys, &col.key_type()),
+            )?;
+        }
+    }
```

Comment:
So we’re going with going column by column rather than having more logic here which explicitly decides which columns and rows to copy based on data it already copied?

Reply:
Don't understand what special logic can we have here. Is there more optimal way of writing data than column by column? I imagine jumping between different columns is costly. 
To clarify, we don't do any reads for keys in combine_keys, and we don't do any writes get_keys_from_store

Reply:
I do understand that it is a more straightforward way, a kind of self-explanatory code that does what we expect in the order in which we expect it*, but I don't feel that it is more structured. In the end, both approaches will have the same code, I mean basically the same lines, there are hardly two ways of reading and writing the same columns from and to our db. But I try to separate logically different stuff, which for me is bringing structure. And it doesn't impact performance in a negative way.
Also, I don't want the information about how cold column's keys are formed to be stored in the file specific to cold storage in non-obvious way. I want to manually form as little keys as possible. 
*As my aproach is not straigthforward, I should add more docs

Reply:
The difference is that you don’t need to collect keys.  You can go entity by entity and deal with each key as you read it.  Ultimately though I don’t have strong feelings.

Reply:
I would like to stick to this approach.